infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -131,60 +142,126 @@ public LogUnitServer(ServerContext serverContext) {
-    public void handleHeadRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+    public void handleTrimMarkRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        log.debug("handleTrimMarkRequest: received a trim mark request {}", msg);
         r.sendResponse(ctx, msg, CorfuMsgType.TRIM_MARK_RESPONSE.payloadMsg(streamLog.getTrimMark()));
     }
 
+    /**
+     * A helper function that maps an exception to the appropriate response message.
+     */
+    void handleException(Throwable ex, ChannelHandlerContext ctx, CorfuPayloadMsg msg, IServerRouter r) {
+        if (ex.getCause() instanceof WrongEpochException) {
+            WrongEpochException wee = (WrongEpochException) ex.getCause();
+            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, wee.getCorrectEpoch()));
+        } else if (ex.getCause() instanceof OverwriteException) {
+            OverwriteException owe = (OverwriteException) ex.getCause();
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE
+                    .payloadMsg(owe.getOverWriteCause().getId()));
+        } else if (ex.getCause() instanceof DataOutrankedException) {
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
+        } else if (ex.getCause() instanceof ValueAdoptedException) {
+            ValueAdoptedException vae = (ValueAdoptedException) ex.getCause();
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(vae.getReadResponse()));
+        } else if (ex.getCause() instanceof TrimmedException) {
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_TRIMMED.msg());
+        } else {
+            throw new LogUnitException(ex);
+        }
+    }
+
     /**
      * Service an incoming write request.
      */
     @ServerHandler(type = CorfuMsgType.WRITE)
     public void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
         log.debug("log write: global: {}, streams: {}", msg.getPayload().getToken(),
                 msg.getPayload().getData().getBackpointerMap());
-
-        try {
-            LogData logData = (LogData) msg.getPayload().getData();
-            logData.setEpoch(msg.getEpoch());
-            dataCache.put(msg.getPayload().getGlobalAddress(), logData);
-            r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
-
-        } catch (OverwriteException ex) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE.payloadMsg(ex.getOverWriteCause().getId()));
-        } catch (DataOutrankedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
-        } catch (ValueAdoptedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(e
-                    .getReadResponse()));
-        }
+        LogData logData = (LogData) msg.getPayload().getData();
+
+        batchWriter.addTask(WRITE, msg)
+                .thenRun(() -> {
```

Comment:
I think using `thenRun()` here has a performance issue. `thenRun()` will use the same thread as `addTask()`, which means if you print the thread name here, you will see it is using the `BatchProcessor` thread to do the following operations (dataCache.put and sendResponse), which is not intended. So I believe the right way to do this is to use `thenRunAsync()` and pass this log unit server's `executor` as the second parameter for `thenRunAsync()`. What do you think?

Reply:
The idea is to be non-blocking. 

Based on my understanding, the LogUnit thread can complete the task (i.e. executes `CompletableFuture::thenRun`) if the future completed "right away", otherwise the LogUnit thread will not block to wait for completion. In the case where the future is not completed "right away", then the completing thread will execute thenRun (i.e. the BatchProcessor thread). Since the post-task work is **minimal** and also **non-blocking** then I think its fine for the BatchProcessor to do it.

Reply:
More info. 

Ran a test where some tasks execute right away. You can see two of the tasks were accepted by the LogUnit threads (i.e. completed right away) and the two others dont. 

o.c.i.LogUnitServer | write: handler thread LogUnit-1
**o.c.i.LogUnitServer | write: Accepting thread LogUnit-1**
 o.c.i.LogUnitServer | write: handler thread LogUnit-4
**o.c.i.LogUnitServer | write: Accepting thread LogUnit-4**
o.c.i.LogUnitServer | write: handler thread LogUnit-7
o.c.i.LogUnitServer | write: Accepting thread LogUnit-BatchProcessor-0
o.c.i.LogUnitServer | write: handler thread LogUnit-2
o.c.i.LogUnitServer | write: Accepting thread LogUnit-BatchProcessor-0
o.c.i.LogUnitServer | write: handler thread LogUnit-5


In another test, I force latency in the writes and you see that all requests are completed by the BatchProcessor. 

o.c.i.LogUnitServer | write: handler thread LogUnit-1
o.c.i.LogUnitServer | write: handler thread LogUnit-2
o.c.i.LogUnitServer | write: handler thread LogUnit-3
o.c.i.LogUnitServer | write: handler thread LogUnit-4
o.c.i.LogUnitServer | write: Accepting thread LogUnit-BatchProcessor-0
o.c.i.LogUnitServer | write: Accepting thread LogUnit-BatchProcessor-0
o.c.i.LogUnitServer | write: Accepting thread LogUnit-BatchProcessor-0
o.c.i.LogUnitServer | write: Accepting thread LogUnit-BatchProcessor-0

Reply:
This might be an issue under high cache contention, or on many cache misses where the BatchProcessor thread competes with the LogUnit server threads.

Reply:
Yes, the  post-task work includes cache and Netty operations. Cache operation may have lock contention and Netty operations may have serialization costs. My point here is batchProcessor thread should be focused on handling log operations, and let the server's executor do all the extra works, which makes more sense since the server executor is meant to do such work, otherwise the single threaded batchProcessor will be overloaded.

Reply:
The netty operations are non-blocking and serialization happens via the netty worker thread, not the LogUnit handler threads, or the BatchProcessor.

At the moment its not an obvious choice to me. Note that the cache operations are "non-blocking" (i.e. since I removed the Writer). But there might be an issue when many of the reads result in cache misses and the loading mechanism starts locking many of the buckets, which can block the BatchProcessor thread. Well if that happens and the LogUnit threads are busy competing on "reads", then whose adding work to the BatchProcessor? 

In the case of no-contention,  why force the overhead of dispatching the work to another thread pool, if its not necessary. That trade-off is not obvious to me. Actually, that's one of the reasons why CompletableFuture tries to run on the same thread if the operation completes right away (i.e. it can be an optimization).